src/parthenon_arrays.hpp

```diff
@@ -0,0 +1,315 @@
+                 Kokkos::ALL(),slc,Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto SliceD(index_pair_t slc, std::integral_constant<int,1>) const {
+    return Slice(SLC0,SLC0,SLC0,Kokkos::ALL(),Kokkos::ALL(),slc);
+  }
+  #undef SLC0
+
+  // AthenaArray.InitWithShallowSlice(src,dim,indx,nvar)
+  // translates into auto dest = src.SliceD<dim>(std::make_pair(indx,indx+nvar))
+  template<std::size_t N = 6>
+  auto SliceD(index_pair_t slc,
+              std::integral_constant<int,N> ic =
+              std::integral_constant<int,N>{}) const {
+    return SliceD(slc, ic);
+  }
+
+  // AthenaArray.InitWithShallowSlice(src,dim,indx,nvar)
+  // translates into auto dest = src.SliceD<dim>(indx,nvar)
+  template<std::size_t N = 6>
+  auto SliceD(const int indx, const int nvar,
+              std::integral_constant<int,N> ic =
+              std::integral_constant<int,N>{}) {
+    return SliceD(std::make_pair(indx,indx+nvar),ic);
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(int i) const {
+    return Kokkos::subview(d6d_,i,
+                           Kokkos::ALL(),Kokkos::ALL(),
+                           Kokkos::ALL(),Kokkos::ALL(),Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(int j, int i) const {
+    return Kokkos::subview(d6d_,j,i,
+                           Kokkos::ALL(),Kokkos::ALL(),
+                           Kokkos::ALL(),Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(int k, int j, int i) const {
+    return Kokkos::subview(d6d_,k,j,i,
+                           Kokkos::ALL(),Kokkos::ALL(),Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(int n, int k, int j, int i) const {
+    return Kokkos::subview(d6d_,n,k,j,i,
+                           Kokkos::ALL(),Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(int m, int n, int k, int j, int i) const {
+    return Kokkos::subview(d6d_,m,n,k,j,i,
+                           Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(int l, int m, int n, int k, int j, int i) const {
+    return Kokkos::subview(d6d_,l,m,n,k,j,i); // 0d view
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get() const {
+    return d6d_;
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(std::integral_constant<int,6>) const {
+    return Get();
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(std::integral_constant<int,5>) const {
+    return Get(0);
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(std::integral_constant<int,4>) const {
+    return Get(0,0);
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(std::integral_constant<int,3>) const {
+    return Get(0,0,0);
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(std::integral_constant<int,2>) const {
+    return Get(0,0,0,0);
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(std::integral_constant<int,1>) const {
+    return Get(0,0,0,0,0);
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(std::integral_constant<int,0>) const {
+    return Get(0,0,0,0,0,0);
+  }
+  template<std::size_t N = 6>
+  KOKKOS_INLINE_FUNCTION
+  auto Get(std::integral_constant<int,N> ic =
```

Comment:
Is there any reason ic cannot be a const reference? Again, I'm a bit out of my depth.

Reply:
This can be const reference, and I changed it. I'm not sure if it does any good, in this case, since the default argument is pretty much always going to be called, and thus constructed, which is ok, since `std::integral_constant` is essentially an empty struct. It's the template-metaprogramming equivalent of an `int`.

The input parameter into this function is just a trick to get the template specialization to work correctly. I.e., the function is intended to be called as, e.g., `Get<3>();`.